iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java

```diff
@@ -505,19 +512,82 @@ private void handleReplaceColumns(org.apache.hadoop.hive.metastore.api.Table hms
     }
 
     for (FieldSchema updatedCol : schemaDifference.getTypeChanged()) {
-      Type newType = HiveSchemaUtil.convert(TypeInfoUtils.getTypeInfoFromTypeString(updatedCol.getType()));
-      if (!(newType instanceof Type.PrimitiveType)) {
-        throw new MetaException(String.format("Cannot promote type of column: '%s' to a non-primitive type: %s.",
-            updatedCol.getName(), newType));
-      }
-      updateSchema.updateColumn(updatedCol.getName(), (Type.PrimitiveType) newType, updatedCol.getComment());
+      updateSchema.updateColumn(updatedCol.getName(), getPrimitiveTypeOrThrow(updatedCol), updatedCol.getComment());
     }
 
     for (FieldSchema updatedCol : schemaDifference.getCommentChanged()) {
       updateSchema.updateColumnDoc(updatedCol.getName(), updatedCol.getComment());
     }
   }
 
+  private void handleChangeColumn(org.apache.hadoop.hive.metastore.api.Table hmsTable) throws MetaException {
+    List<FieldSchema> hmsCols = hmsTable.getSd().getCols();
+    List<FieldSchema> icebergCols = HiveSchemaUtil.convert(icebergTable.schema());
+    // compute schema difference for renames, type/comment changes
+    HiveSchemaUtil.SchemaDifference schemaDifference = HiveSchemaUtil.getSchemaDiff(hmsCols, icebergCols, true);
+    // check column reorder (which could happen even in the absence of any rename, type or comment change)
+    Map<String, String> renameMapping = ImmutableMap.of();
+    if (!schemaDifference.getMissingFromSecond().isEmpty()) {
+      renameMapping = ImmutableMap.of(
+          schemaDifference.getMissingFromSecond().get(0).getName(),
+          schemaDifference.getMissingFromFirst().get(0).getName());
+    }
+    Pair<String, Optional<String>> outOfOrder = HiveSchemaUtil.getFirstOutOfOrderColumn(hmsCols, icebergCols,
+        renameMapping);
+
+    if (!schemaDifference.isEmpty() || outOfOrder != null) {
+      updateSchema = icebergTable.updateSchema();
+    } else {
+      // we should get here if the user restated the exactly the existing column in the CHANGE COLUMN command
```

Comment:
Please fix the comment, I do not get it 😄

Reply:
Or fix me 😄

Reply:
If the comment is not clear to you, it needs to be fixed :) Will do it

Reply:
Updated the comment, let me know if this clarifies it.